### PR TITLE
add --disable-auto-connect flag

### PR DESCRIPTION
**Pull request by nonsense**
This PR is adding a `--disable-auto-connect` flag, so that we can run Swarm without `hive` discovery and reproduce the same connections between deployments, without relying on the non-deterministic `SuggestPeer` functionality.

Next step would be to add tools to extract and apply connections from running deployments (a very early attempt at https://github.com/ethersphere/swarm/pull/1183) and to generate snapshots out of them so that we can have more determinism in our tests.

### CODE REVIEW

File: network/hive.go [hunk 1]
```diff
@@ -138,14 +137,17 @@ func (h *Hive) connect() {
 			continue
 		}
 
-		log.Trace(fmt.Sprintf("%08x hive connect() suggested %08x", h.BaseAddr()[:4], addr.Address()[:4]))
-		under, err := enode.ParseV4(string(addr.Under()))
-		if err != nil {
-			log.Warn(fmt.Sprintf("%08x unable to connect to bee %08x: invalid node URL: %v", h.BaseAddr()[:4], addr.Address()[:4], err))
-			continue
+		if h.Discovery {
```
Comment (nonsense) [inline network/hive.go]: Note that if `h.addPeer` is not under the `if h.Discovery` clause, we were still making connections due to `hive` even if `h.Discovery` was set to `false`...

**Review by janos — DISMISSED** [review]

**Review by skylenet — DISMISSED** [review]

File: network/hive.go [hunk 2]
```diff
@@ -83,7 +83,7 @@ func NewHive(params *HiveParams, kad *Kademlia, store state.Store) *Hive {
 // server is used to connect to a peer based on its NodeID or enode URL
 // these are called on the p2p.Server which runs on the node
 func (h *Hive) Start(server *p2p.Server) error {
-	log.Info("Starting hive", "baseaddr", fmt.Sprintf("%x", h.BaseAddr()[:4]))
+	log.Info("Starting hive", "baseaddr", fmt.Sprintf("%x", h.BaseAddr()[:4]), "discovery", h.Discovery)
```
Comment (zelig) [inline network/hive.go]: What you want is not start the connect loop on  100 below if `Discovery` is set to false

File: network/hive.go [hunk 3]
```diff
@@ -129,7 +129,6 @@ func (h *Hive) Stop() error {
 // as well as advertises saturation depth if needed
 func (h *Hive) connect() {
 	for range h.ticker.C {
-
 		addr, depth, changed := h.SuggestPeer()
 		if h.Discovery && changed {
```
Comment (zelig) [inline network/hive.go]: lets remove checking of Discovery here

File: network/hive.go (hunk 1 above)
Comment (zelig) [reply network/hive.go]: lets remove checking of Discovery here, simply the connect loop should not be started

**Review by zelig — CHANGES_REQUESTED** [review]
the flag should cause  `hive.Start` not to start the `connect` loop

File: network/hive.go (hunk 2 above)
Comment (nonsense) [reply network/hive.go]: I consider that, but the flag being used on line 133 confused me. I think the logic would still be the same if we don't start the `connect loop` as it is now, so the tests would still fail, so I have to investigate that first.

File: network/simulations/discovery/discovery_test.go [hunk 4]
```diff
@@ -147,7 +147,7 @@ func TestDiscoverySimulationSimAdapter(t *testing.T) {
 	testDiscoverySimulationSimAdapter(t, *nodeCount, *initCount)
 }
 
-func TestDiscoveryPersistenceSimulationSimAdapter(t *testing.T) {
+func XTestDiscoveryPersistenceSimulationSimAdapter(t *testing.T) {
```
Comment (nonsense) [inline network/simulations/discovery/discovery_test.go]: @zelig could you maybe add a few comments on what this test is supposed to test?

I am also not sure how the vars `persistenceEnabled` and `discoveryEnabled` are supposed to work - what if tests are run concurrently?

It seems to me that this test is starting a network and connecting nodes in a chain, and then waiting for the Kademlia table to be setup by `hive`. At this point it stops the nodes, stops discovery and starts them back up, expecting for the Kademlia table to be setup `without discovery`.

If this is correct, it seems like we are changing the meaning of the `NoDiscovery` flag - previously it was just preventing `subPeerMsg` message exchanges, and not preventing actual connections, whereas now if we set the `NoDiscovery` flag to true, we don't want `hive` to trigger any connections what-so-ever.

Therefore if my assumption on this test is correct, it is expected for it to fail after the change.

What do you think?

Comment (zelig) [reply network/simulations/discovery/discovery_test.go]: well you seemed to want to redefine this flag to not use the address book that is saved just allow manual (or snapshot driven) connection. 

This test is supposed to test that peers persist across sessions, i.e., that the addressbook is saved and used when bootstrapping connectivity

**Review by zelig — COMMENTED** [review]
address book persistence test should not be removed.
Maybe we need two flags: `discoveryDisabled` and `autoconnectDisabled`.
i am not sure of `persistanceEnabled` , cannot see the usecase of disabling it

**Comment by nonsense** [comment]
Yes, the definition of `discovery` is changing, and we should decide what to do with it.

We obviously need a `discovery` flag, which pretty much disables both message exchange in `hive` as well as automatic connections, because we want to be able to deterministically setup networks.

I am not sure if we need an `autoconnect` flag - it seems to be used only to test persistence, and we can probably test persistence without this flag. Furthermore we want the Kademlia table to be mostly constructed while preferring peers we had connections to before, rather than just kicking in the Discovery mechanism and randomly getting new ones.

File: network/simulations/discovery/discovery_test.go (hunk 4 above)
Comment (nonsense) [reply network/simulations/discovery/discovery_test.go]: BTW the addressbook persistence is broken for a long time - if there are peers that no longer exist, we never clear them out, and we keep on trying to connect to them indefinitely - something we tried fixing with @homotopycolimit at some point, but didn't work nicely as a `quick hack`.

**Comment by nonsense** [comment]
Address book persistence in my opinion doesn't work well, so should be getting an issue for itself, and is probably out of scope for this PR. If there are peers in it that are not existent anymore, a node will try to connect to them indefinitely. I think it is more important to be able to deterministically build networks right now, so that we can improve our overall testing - something that @acud is having pain with in `pull sync` AFAICT seeing the new issues.

File: network/simulations/discovery/discovery_test.go (hunk 4 above)
Comment (acud) [reply network/simulations/discovery/discovery_test.go]: @nonsense this test was my first coding task on swarm so i take responsibility :D

Discovery should indeed not impact on whether we are connecting to peer from our address book, so I don't see why this should impact the way we connect to existing known peers, and I'm not sure that disabling the test would help this fact at all

**Review by acud — COMMENTED** [review]
@nonsense let's re-approach this.
do you want to block the possibility of getting notified about new peers from your peers or would you like to disable address book persistence?

**Comment by nonsense** [comment]
```
@nonsense let's re-approach this.
do you want to block the possibility of getting notified about new peers from your peers or would you like to disable address book persistence?
```

@acud this issue/PR is not at all about any of this. I don't see `address book persistence` having to do anything with `hive` being disabled and disabling hive seems to impact its tests - I actually think the test is wrong, because it expects that we will be connecting to peers when `NoDiscovery` is set to false, whereas the only thing we do with the address book is load it in Kademlia.

So the question is do we want to couple `address book loading` with `NoDiscovery` - I'd say no.

Ultimately I just want to be able to start Swarm without having `hive` create connections to peers, so that we can have deterministic networks.

**Comment by nonsense** [comment]
@zelig @acud we spoke about this on a meeting 2-3 weeks ago, and I claimed that there is no way to create a deterministic network with Kademlia - I think this PR mostly shows that, because there is no way to disable hive and to stop it from making connections.

So how do you want to complete this?

I think the test that checks for persistence is wrong, because it checks for connectivity and not for saving/loading of peers. In general I think persistence should be improved as a larger effort, since it doesn't clear up old peers.

Bottom line - I don't think the `Discovery` flag (which cannot be set by a command line argument right now) is useful at all in its current state, therefore I am changing its behaviour. It seems to be used only for the persistence test.

Having the possibility to create deterministic networks is quite important for all our efforts right now, so I'd suggest we go forward with this, and not wait for the Kademlia refactor we have on the roadmap. I think it'd be useful to have reproducible tests sooner rather than later.

File: network/simulations/discovery/discovery_test.go (hunk 4 above)
Comment (zelig) [reply network/simulations/discovery/discovery_test.go]: In fact, it makes more sense to have `--no-hive` to mean no connect loop. So you could simply fix this test by wrapping hive service in a struct and redefine its Protocols function to return empty. That way you can get the hive service with connect loop but no protocol running, so you can safely test bootstrapping from addressbook.

**Comment by acud** [comment]
I think the basic misunderstanding here is the discovery flag. This actually has `nil` to do with discovery. Discovery is the act of gossiping peers to your peers and this is what the flag should control. 

The persistence test semantics to this flag in this case is _correct_, since the content of the persisted address book should be bootstrapped to the node (or at least intercepted with a hook and asserted correctly).

I agree that address book persistence should not be related to hive. If currently the address book content is loaded through hive then this is incorrect and should be done through the p2p server somehow.

**Review by acud — DISMISSED** [review]

File: network/simulations/discovery/discovery_test.go (hunk 4 above)
Comment (nonsense) [reply network/simulations/discovery/discovery_test.go]: I'd rather not redefine the `Protocols` function to test persistence of the kademlia address book, this seems like a very ugly hack.

With this PR `--no-hive-discovery` does not run the `connect()` loop.

Let's talk about the address book as part of the Kademlia epic, this is out of scope for this PR. Both me and Elad seem to think that similar to how bootnodes are connected to outside of `hive`, it makes sense to do something similar for the address book.

**Review by janos — DISMISSED** [review]

**Comment by zelig** [comment]
1. I am in favour of getting rid of the current interpretation of hive discovery (whether to gossip peer information via the hive protocol) because i dont see a need for disabling hive protocol apart from the persistance test. 

2. The persistance test is supposed to test if the hive correctly remembers and reuses the peers in the address book  accross sessions when you bootstrap connectivity for a node. If hive protocol is gossiping peer info, this cannot be tested easily, so i recommend simply wrapping the hive service in a struct and redefine its Protocols function to return empty. That way you can get the hive service with connect loop but no protocol running, so you can safely test bootstrapping from addressbook without allowing such configuration in production code. This is not a lot of work. 

3. In https://github.com/ethersphere/swarm/issues/1535 There has been a new requirement with regards to bootstrapping: to give preference to peers we were last connected (if they exist) over other peers (even if they are freshly being gossipped). While this needs to be further specified, it is likely to replace the persistance test described in 2. so ultimately we will not need to support a case with running connect loop with no hive protocol. But I really prefer fixing tests when you change something, rather than disable or delete them just cos they will need to be reworked at some future time. 

4. As for the flag, I recommend renaming it to `-disable-auto-connect` to avoid ambiguity and simply have it not call `hive.Start` or not start the `go connect()` in the hive start. 

5. As for bootstrapping from snapshot and saving a snapshot i guess there are https://github.com/ethersphere/swarm/issues/1584 and https://github.com/ethersphere/swarm/issues/1585 now.

**Comment by nonsense** [comment]
@zelig about 2: 

I'd rather not redefine the `Protocols` function to test persistence of the Kademlia address book, this seems like a very ugly hack. This feels like the completely wrong level to test something like address book persistence. It is not a matter of a lot of work, or not. Furthermore currently we seem to only `Register` the loaded peers, and not connect to them. This is not really reproducible, because the idea is that you connect to peers that you were previously connected to, whereas now Hive is free to override that `loaded peers` list and connect to other `suggested peers`.

File: api/config.go [hunk 5]
```diff
@@ -53,32 +53,32 @@ type Config struct {
...
+	Enode                  *enode.Node `toml:"-"`
+	NetworkID              uint64
+	SwapEnabled            bool
+	SyncEnabled            bool
+	SyncingSkipCheck       bool
+	DeliverySkipCheck      bool
+	MaxStreamPeerServers   int
+	LightNodeEnabled       bool
+	BootnodeMode           bool
+	HiveDisableAutoConnect bool
```
Comment (zelig) [inline api/config.go]: no need for Hive prefix

File: network/simulation/kademlia_test.go [hunk 6]
```diff
@@ -130,7 +130,7 @@ func createSimServiceMap(discovery bool) map[string]ServiceFunc {
 		"bzz": func(ctx *adapters.ServiceContext, b *sync.Map) (node.Service, func(), error) {
 			addr := network.NewAddr(ctx.Config.Node())
 			hp := network.NewHiveParams()
-			hp.Discovery = discovery
+			hp.AutoConnect = discovery
```
Comment (zelig) [inline network/simulation/kademlia_test.go]: please change the variable name too

File: network/simulations/discovery/discovery_test.go [hunk 7]
```diff
@@ -186,26 +178,6 @@ func testDiscoverySimulation(t *testing.T, nodes, conns int, adapter adapters.No
 	t.Logf("Setup: %s, shutdown: %s", result.StartedAt.Sub(startedAt), finishedAt.Sub(result.FinishedAt))
 }
 
-func testDiscoveryPersistenceSimulation(t *testing.T, nodes, conns int, adapter adapters.NodeAdapter) map[int][]byte {
```
Comment (zelig) [inline network/simulations/discovery/discovery_test.go]: I disagree with removing this test when you can easily fix it. Please

File: network/simulations/discovery/discovery_test.go [hunk 8]
```diff
@@ -513,9 +328,9 @@ func newService(ctx *adapters.ServiceContext) (node.Service, error) {
 	kad := network.NewKademlia(addr.Over(), kp)
 	hp := network.NewHiveParams()
 	hp.KeepAliveInterval = time.Duration(200) * time.Millisecond
-	hp.Discovery = discoveryEnabled
+	hp.AutoConnect = discoveryEnabled
```
Comment (zelig) [inline network/simulations/discovery/discovery_test.go]: please change the variable name too. otherwise we accumulate technical debt

File: network/simulations/overlay.go [hunk 9]
```diff
@@ -94,7 +94,7 @@ func (s *Simulation) NewService(ctx *adapters.ServiceContext) (node.Service, err
 	kp.RetryInterval = 1000000
 	kad := network.NewKademlia(addr.Over(), kp)
 	hp := network.NewHiveParams()
-	hp.Discovery = !*noDiscovery
+	hp.AutoConnect = !*noDiscovery
```
Comment (zelig) [inline network/simulations/overlay.go]: same

**Review by zelig — CHANGES_REQUESTED** [review]
minor plus I think address book persistance simulation should be restored and adapted

**Review by zelig — APPROVED** [review]
allright so you are only adding a new flag. Fine.

**Review by skylenet — APPROVED** [review]
